trace_decoder/src/decoding.rs

```diff
@@ -32,11 +34,120 @@ use crate::{
+/// arise when parsing, validating, or otherwise processing the trace data of
+/// blockchain blocks. It could include issues like malformed trace data,
+/// inconsistencies found during processing, or any other condition that
+/// prevents successful completion of the trace processing task.
+#[derive(Debug)]
+pub struct TraceParsingError {
+    block_num: Option<U256>,
+    block_chain_id: Option<U256>,
+    txn_idx: Option<usize>,
+    addr: Option<Address>,
+    h_addr: Option<H256>,
+    slot: Option<U512>,
+    slot_value: Option<U512>,
+    reason: TraceParsingErrorReason, // The original error type
+}
+
+impl std::fmt::Display for TraceParsingError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Error processing trace: {}", self.reason)?;
+
+        if let Some(block_num) = self.block_num {
+            write!(f, "\nBlock num: {}", block_num)?;
+        }
+        if let Some(block_chain_id) = self.block_chain_id {
+            write!(f, "\nBlock chain id: {}", block_chain_id)?;
+        }
+        if let Some(txn_idx) = self.txn_idx {
+            write!(f, "\nTxn idx: {}", txn_idx)?;
+        }
+        if let Some(addr) = self.addr {
+            write!(f, "\nAddress: {}", addr)?;
+        }
+        if let Some(h_addr) = self.h_addr {
+            write!(f, "\nHashed address: {}", h_addr)?;
+        }
+        if let Some(slot) = self.slot {
+            write!(f, "\nSlot: {:x}", slot)?;
+        }
+        if let Some(slot_value) = self.slot_value {
+            write!(f, "\nSlot value: {:x}", slot_value)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl std::error::Error for TraceParsingError {}
+
+impl TraceParsingError {
+    /// Function to create a new TraceParsingError with mandatory fields
+    pub(crate) fn new(reason: TraceParsingErrorReason) -> Self {
+        Self {
+            block_num: None,
+            block_chain_id: None,
+            txn_idx: None,
+            addr: None,
+            h_addr: None,
+            slot: None,
+            slot_value: None,
+            reason,
+        }
+    }
+
+    /// Builder method to set block_num
+    pub(crate) fn block_num(&mut self, block_num: U256) -> &mut Self {
+        self.block_num = Some(block_num);
+        self
+    }
+
+    /// Builder method to set block_chain_id
+    pub(crate) fn block_chain_id(&mut self, block_chain_id: U256) -> &mut Self {
+        self.block_chain_id = Some(block_chain_id);
+        self
+    }
+
+    /// Builder method to set txn_idx
+    pub fn txn_idx(&mut self, txn_idx: usize) -> &mut Self {
+        self.txn_idx = Some(txn_idx);
+        self
+    }
+
+    /// Builder method to set addr
+    pub fn addr(&mut self, addr: Address) -> &mut Self {
+        self.addr = Some(addr);
+        self
+    }
+
+    /// Builder method to set h_addr
+    pub fn h_addr(&mut self, h_addr: H256) -> &mut Self {
+        self.h_addr = Some(h_addr);
+        self
+    }
```

Comment:
Nit: It might be better to just set `h_addr` and then print out the hashed address along with the `addr` instead of storing it separately. If we ever have the address, then we always have the hashed address.

Reply:
but here https://github.com/0xPolygonZero/zk_evm/pull/148/files#diff-4bf5db436239b600423f322fb2abf7364bb676107d1bac9526951585a777c16fR392 we set just the h_addr

Reply:
Oh you mean like this?
```rust
TraceParsingErrorReason::MissingAccountStorageTrie(hashed_addr),
```

Yeah that's a fair point. When I created these errors with `this_error`, I couldn't find a way to call functions inside the `#[error(...)]` annotations, and I didn't think I could hash a stored value. Looking at it now however, it looks like there might be an undocumented feature to do so:

```rust
pub enum Error {
    #[error("first letter must be lowercase but was {:?}", first_char(.0))]
    WrongCase(String),
```

Love dtolnay's work, but I wish he did better docs...

Anyways yeah this is just a nit, and I'm good either way.